File: evaluation/S0/python/Medium/842-next-greater-element-ii.py
# Source: https://leetcode.com/problems/next-greater-element-ii/   |   Difficulty: Medium
#
# Problem Description:
# Given a circular integer array nums (i.e., the next element of nums[nums.length - 1] is nums[0]), return the next greater number for every element in nums.
#
# The next greater number of a number x is the first greater number to its traversing-order next in the array, which means you could search circularly to find its next greater number. If it doesn't exist, return -1 for this number.
#
# Example:
# Input: nums = [1,2,1]
# Output: [2,-1,2]
# Explanation: The first 1's next greater number is 2; 
# The number 2 can't find next greater number. 
# The second 1's next greater number needs to search circularly, which is also 2.
#
# Constraints:
# 1 <= nums.length <= 104
# 	-109 <= nums[i] <= 109
#

class Solution(object):
    def nextGreaterElements(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        res = [-1] * n
        stack = []
        
        for i in range(2 * n):
            num = nums[i % n]
            while stack and nums[stack[-1]] < num:
                index = stack.pop()
                res[index] = num
            if i < n:
                stack.append(i)
        
        return res

File: evaluation/S0/java/Medium/842-next-greater-element-ii.java
// Source: https://leetcode.com/problems/next-greater-element-ii/   |   Difficulty: Medium
//
// Problem Description:
// Given a circular integer array nums (i.e., the next element of nums[nums.length - 1] is nums[0]), return the next greater number for every element in nums.
//
// The next greater number of a number x is the first greater number to its traversing-order next in the array, which means you could search circularly to find its next greater number. If it doesn't exist, return -1 for this number.
//
// Example:
// Input: nums = [1,2,1]
// Output: [2,-1,2]
// Explanation: The first 1's next greater number is 2; 
// The number 2 can't find next greater number. 
// The second 1's next greater number needs to search circularly, which is also 2.
//
// Constraints:
// 1 <= nums.length <= 104
// 	-109 <= nums[i] <= 109
//

class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] result = new int[n];
        Arrays.fill(result, -1);
        Deque<Integer> stack = new ArrayDeque<>();

        for (int i = 0; i < 2 * n; i++) {
            int currentIndex = i % n;
            while (!stack.isEmpty() && nums[stack.peek()] < nums[currentIndex]) {
                int index = stack.pop();
                result[index] = nums[currentIndex];
            }
            if (i < n) {
                stack.push(currentIndex);
            }
        }

        return result;          
    }
}
